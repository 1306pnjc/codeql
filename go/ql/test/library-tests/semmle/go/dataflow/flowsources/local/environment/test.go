package test

import (
	"fmt"
	"github.com/hashicorp/go-envparse"
	"github.com/joho/godotenv"
	"github.com/kelseyhightower/envconfig"
	"os"
)

func osEnvironmentVariables() {
	home := os.Getenv("HOME")

	port, ok := os.LookupEnv("PORT")
	if !ok {
		port = "3000"
	}

	for _, e := range os.Environ() {
		_ = e
	}

	fmt.Printf("HOME: %s\n", home)
	fmt.Printf("PORT: %s\n", port)
}

type ServerConfig struct {
	Port int    `envconfig:"PORT"`
	Host string `envconfig:"HOST"`
}

func envconfigEnvironmentVariables() {
	var cfg ServerConfig
	envconfig.Process("myapp", &cfg)
}

func godotenvEnvironmentVariables() {
	var err error
	var username, greeting string

	users, err := godotenv.Read("user.env")
	if err != nil {
		return
	}

	username := users["USERNAME"]

	greetings, err := godotenv.Unmarshal("HELLO=hello")
	if err != nil {
		return
	}

	greeting := greetings["HELLO"]

	fmt.Printf("%s, %s!\n", greeting, username)
}

func envparseEnvironmentVariables() {
	f := os.OpenRead("file.txt")
	defer f.Close()
	envVars, ok := envparse.Parse(f)

	if !ok {
		return
	}

	fmt.Printf("HOME: %s\n", envVars["HOME"])
}
