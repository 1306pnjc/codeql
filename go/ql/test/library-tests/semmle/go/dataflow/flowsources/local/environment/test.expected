| test.go:12:10:12:26 | call to Getenv |
| test.go:14:2:14:33 | ... := ...[0] |
| test.go:19:20:19:31 | call to Environ |
| test.go:34:29:34:32 | &... |
| test.go:41:2:41:40 | ... := ...[0] |
| test.go:48:2:48:52 | ... := ...[0] |
| test.go:61:2:61:33 | ... := ...[0] |
